server-metrics/src/main/scala/org/http4s/server/metrics/Metrics.scala

```diff
@@ -2,102 +2,183 @@ package org.http4s
-      OptionT {
-        service(req).value.attempt
-          .flatMap(onFinish(req.method, now)(_)
-            .fold[F[Option[Response[F]]]](F.raiseError, F.pure))
+    Kleisli(metricsService[F](serviceMetrics, service)(_))
+  }
+
+  private def metricsService[F[_]: Sync](serviceMetrics: ServiceMetrics, service: HttpService[F])(req: Request[F]): OptionT[F, Response[F]] = {
+    val method = req.method
+    for {
+      now <- OptionT.liftF[F, Long](Sync[F].delay(System.nanoTime()))
+      _ <- OptionT.liftF[F, Unit](Sync[F].delay(serviceMetrics.generalMetrics.active_requests.inc()))
+      resp <- OptionT{
+        service(req)
+          .value
+          .attempt
+          .flatMap(metricsServiceHandler(method, now, serviceMetrics, _))
       }
+    } yield resp
+  }
+
+  private def metricsServiceHandler[F[_]: Sync](method: Method,
+                                                start: Long,
+                                                serviceMetrics: ServiceMetrics,
+                                                e: Either[Throwable, Option[Response[F]]]
+                                               ): F[Option[Response[F]]] = {
+    for {
+      elapsed <- EitherT.liftF[F, Throwable, Long](Sync[F].delay(System.nanoTime() - start))
+      respOpt <- EitherT(e.bitraverse[F, Throwable, Option[Response[F]]](
+        manageServiceErrors(method, elapsed, serviceMetrics).as(_),
+        _.map(manageResponse(method, elapsed, serviceMetrics)).pure[F]
+      ))
+    } yield respOpt
+  }.fold(
+    Sync[F].raiseError[Option[Response[F]]],
+    _.fold(handleUnmatched(serviceMetrics.generalMetrics.active_requests))(handleMatched)
+  ).flatten
+
+  private def manageResponse[F[_]: Sync](
+                                   m: Method,
+                                   elapsed: Long,
+                                   serviceMetrics: ServiceMetrics
+                                 )(response: Response[F]): Response[F] = {
+    val newBody = response.body
+      .onFinalize(
+        incrementCounts(serviceMetrics.generalMetrics.headers_times, elapsed) *>
+        requestMetrics(serviceMetrics.requestTimers, serviceMetrics.generalMetrics.active_requests)(m, elapsed) *>
+        responseMetrics(serviceMetrics.responseTimers, response.status, elapsed)
+      )
+      .handleErrorWith(e =>
+        Stream.eval(incrementCounts(serviceMetrics.generalMetrics.abnormal_terminations, elapsed)) *>
+        Stream.raiseError[Byte](e)
+      )
+    response.copy(body = newBody)
+  }
+
+  private def manageServiceErrors[F[_]: Sync](m: Method, elapsed: Long, serviceMetrics: ServiceMetrics): F[Unit] =
+    requestMetrics(serviceMetrics.requestTimers, serviceMetrics.generalMetrics.active_requests)(m, elapsed) *>
+    incrementCounts(serviceMetrics.responseTimers.resp5xx, elapsed) *>
+    incrementCounts(serviceMetrics.generalMetrics.service_errors, elapsed)
+
+
+  private def handleUnmatched[F[_]: Sync](c: Counter): F[Option[Response[F]]] =
```

Comment:
If we increment 5xx on these, should we increment 4xx on this?  This is the dirty bit about doing metrics as middleware: we don't know what happens outside middleware.  Maybe we're chained to something that recovers.  Maybe the server handles it.  I think this is inconsistent with `manageServiceErrors`, though I don't have a strong feeling which direction to fix it.

Reply:
Yeah. I can increment as it without it we don't match requirement 4

Reply:
I question whether us projecting a response is the right move as well. Metrics on a service should measure what happens with that service. For metrics matching unmatched They should make the service total

Reply:
:+1: I think we just count nones and errors and not double count as 4xx and 5xx. Which is a reversal of point 4 above, but I think correct.